### PR TITLE
Don't resolve and retransform assets in JSRuntime

### PR DESCRIPTION
**Pull request by wbinnssmith**
In collaboration with @mischnic 

Previously, when processing async dependencies in the JSRuntime that do not require loading new bundles, we dynamically created new, synchronous `require`s, which processed these assets again. This means resolving and transforming them (though transformations should have resulted in cache hits) unnecessarily.

This:
* In the case of the non-scope-hoisting packager, results in a runtime that exports `Promise.resolve(module.bundle.root(assetId))` instead of `Promise.resolve(require('./path/to/asset'))`
* In the case of the scope-hoisting packager, does not produce a runtime for these dependencies. Instead, the packager resolves the dependency again and inserts `Promise.resolve(identifierOrInitCall)` in its place.

Test Plan:
* Updated existing tests, including those for these "internalized" assets
* @mischnic added an internalized asset example
* [x] Performance test in a real-world app: Looks like this reduces development build times by about 5%

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 10.93s | +5.14s ⚠️ |
| Cached | 798.00ms | -2.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 109.00ms | +67.00ms ⚠️ |
| dist/legacy/index.6fa78f1c.js | 1.07kb | +0.00b | 2.13s | +826.00ms ⚠️ |
| dist/modern/index.df30fe40.js | 1.07kb | +0.00b | 1.71s | +320.00ms ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 2.24s | +770.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 1.96s | +437.00ms ⚠️ |
| dist/legacy/index.afdbda6b.css | 77.00b | +0.00b | 2.09s | +901.00ms ⚠️ |
| dist/modern/index.ef3960f9.css | 77.00b | +0.00b | 1.69s | +420.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/index.6fa78f1c.js | 1.07kb | +0.00b | 81.00ms | +5.00ms ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 81.00ms | +5.00ms ⚠️ |
| dist/legacy/index.afdbda6b.css | 77.00b | +0.00b | 81.00ms | +5.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 24.56s | -34.00ms |
| Cached | 12.95s | -482.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/NotFound.141e0b29.js | 532.00b | +0.00b | 428.00ms | +142.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 91.00ms | +43.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 479.93kb | +0.00b | 66.00ms | -4.00ms 🚀 |
| dist/PermalinkedComment.a5120a24.js | 4.22kb | +0.00b | 66.00ms | -4.00ms 🚀 |
| dist/UserProfile.95e14af2.js | 1.70kb | +0.00b | 66.00ms | -4.00ms 🚀 |
| dist/NotFound.141e0b29.js | 532.00b | +0.00b | 67.00ms | +11.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 45.00ms | +4.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.28m | -4.10s |
| Cached | 3.06s | -33.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/pdfRenderer.705aa8d6.js | 1.11mb | -3.00b 🚀 | 44.28s | -2.69s 🚀 |
| dist/editorView.2c405cb8.js | 587.39kb | +0.00b | 1.13m | -3.54s 🚀 |
| dist/popup.677c7b55.js | 169.35kb | +0.00b | 44.28s | -2.69s 🚀 |
| dist/Toolbar.c7fe4bad.js | 99.25kb | +0.00b | 47.54s | -3.56s 🚀 |
| dist/media-viewer.52bfe015.js | 72.71kb | +0.00b | 12.38s | -1.57s 🚀 |
| dist/card.93c390cd.js | 53.76kb | +0.00b | 1.25m | +1.05m ⚠️ |
| dist/card.8a18675b.js | 51.51kb | +0.00b | 45.29s | -3.15s 🚀 |
| dist/Modal.1933f1f9.js | 39.02kb | +0.00b | 4.39s | -267.00ms 🚀 |
| dist/component.f31cba5f.js | 30.87kb | +0.00b | 2.00s | -167.00ms 🚀 |
| dist/component.c8d4c7aa.js | 22.40kb | +0.00b | 4.39s | -275.00ms 🚀 |
| dist/smartMediaEditor.73dadc38.js | 16.79kb | +0.00b | 46.30s | -3.45s 🚀 |
| dist/dropzone.1a763721.js | 16.03kb | +0.00b | 1.25m | +27.16s ⚠️ |
| dist/workerHasher.1a860446.js | 11.90kb | +0.00b | 8.53s | -1.14m 🚀 |
| dist/component.aaadd915.js | 6.27kb | +0.00b | 6.14s | +1.11s ⚠️ |
| dist/card.3fd8a427.js | 5.71kb | +0.00b | 1.25m | +1.02m ⚠️ |
| dist/media-viewer.b91b6ee1.js | 4.01kb | +0.00b | 12.47s | -1.50s 🚀 |
| dist/EmojiPickerComponent.53446e07.js | 3.68kb | +0.00b | 8.49s | -2.86s 🚀 |
| dist/index.cfd8f78a.css | 3.46kb | +0.00b | 49.43s | -2.72s 🚀 |
| dist/dropzone.dcab7b1b.js | 3.44kb | +0.00b | 22.51s | -22.31s 🚀 |
| dist/Modal.d8614cea.js | 3.15kb | +0.00b | 2.00s | -160.00ms 🚀 |
| dist/clipboard.a5615679.js | 2.97kb | +0.00b | 44.72s | -2.96s 🚀 |
| dist/16.4fc0f868.js | 2.49kb | +0.00b | 2.50s | +217.00ms ⚠️ |
| dist/ResourcedEmojiComponent.87af091a.js | 2.15kb | +0.00b | 37.43s | +26.07s ⚠️ |
| dist/card.bfb3f5f8.js | 2.15kb | +0.00b | 22.41s | -1.94s 🚀 |
| dist/feedback.89b37c2d.js | 1.86kb | +0.00b | 6.00s | -4.18s 🚀 |
| dist/browser.dc14c9f2.js | 1.82kb | +0.00b | 44.77s | -32.27s 🚀 |
| dist/16.fb8534c0.js | 1.79kb | +0.00b | 6.94s | +1.66s ⚠️ |
| dist/workerHasher.209d7f04.js | 1.75kb | +0.00b | 10.50s | -1.11m 🚀 |
| dist/workerHasher.b62cf8f4.js | 1.75kb | +0.00b | 46.11s | -3.48s 🚀 |
| dist/16.85df7a6e.js | 1.51kb | +0.00b | 4.59s | -273.00ms 🚀 |
| dist/16.34fd12e7.js | 1.46kb | +0.00b | 6.16s | +1.13s ⚠️ |
| dist/16.ff3f3afd.js | 1.46kb | +0.00b | 4.59s | -247.00ms 🚀 |
| dist/16.55d1de3e.js | 1.45kb | +0.00b | 2.00s | -161.00ms 🚀 |
| dist/16.9689f28f.js | 1.44kb | +0.00b | 5.24s | -2.30s 🚀 |
| dist/16.ac531af9.js | 1.40kb | +0.00b | 5.24s | -2.30s 🚀 |
| dist/expand.a88194c2.js | 1.38kb | +0.00b | 8.13s | -2.04s 🚀 |
| dist/16.341ad0e1.js | 1.34kb | +0.00b | 2.00s | -163.00ms 🚀 |
| dist/16.0ef0d0ed.js | 1.31kb | +0.00b | 5.24s | -2.30s 🚀 |
| dist/media-card-analytics-error-boundary.9f8e76f5.js | 1.30kb | +0.00b | 22.41s | -1.94s 🚀 |
| dist/16.8015c4e6.js | 1.27kb | +0.00b | 2.00s | -163.00ms 🚀 |
| dist/16.30604232.js | 1.26kb | +0.00b | 4.59s | -247.00ms 🚀 |
| dist/16.a86c2eba.js | 1.26kb | +0.00b | 4.59s | -247.00ms 🚀 |
| dist/16.8e47d46b.js | 1.26kb | +0.00b | 4.59s | -276.00ms 🚀 |
| dist/16.867d534c.js | 1.26kb | +0.00b | 6.16s | +1.13s ⚠️ |
| dist/component.0a97dc28.js | 1.23kb | +0.00b | 2.50s | +217.00ms ⚠️ |
| dist/16.a43add3b.js | 1.23kb | +0.00b | 2.47s | +185.00ms ⚠️ |
| dist/list.36f0ec8a.js | 1.18kb | +0.00b | 5.57s | -32.31s 🚀 |
| dist/media-picker-analytics-error-boundary.807445d0.js | 1.05kb | +0.00b | 44.72s | -2.96s 🚀 |
| dist/media-card-analytics-error-boundary.5602bdf2.js | 1.05kb | +0.00b | 1.25m | +1.05m ⚠️ |
| dist/media-card-analytics-error-boundary.fa8679e7.js | 1.05kb | +0.00b | 45.29s | -3.15s 🚀 |
| dist/media-viewer-analytics-error-boundary.bc280d38.js | 995.00b | +0.00b | 12.47s | -1.52s 🚀 |
| dist/simpleHasher.03c3cf37.js | 755.00b | +0.00b | 8.49s | -1.14m 🚀 |
| dist/simpleHasher.566203bd.js | 755.00b | +0.00b | 46.12s | -3.48s 🚀 |
| dist/index.html | 119.00b | +0.00b | 4.18s | +1.18s ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.6e469c30.js | 2.32mb | +0.00b | 583.00ms | -32.00ms 🚀 |
| dist/Modal.8e9ccd89.js | 38.79kb | +0.00b | 520.00ms | +63.00ms ⚠️ |
| dist/component.f31cba5f.js | 30.87kb | +0.00b | 507.00ms | +67.00ms ⚠️ |
| dist/component.c8d4c7aa.js | 22.40kb | +0.00b | 521.00ms | +63.00ms ⚠️ |
| dist/js.45b69177.js | 16.55kb | +0.00b | 521.00ms | +63.00ms ⚠️ |
| dist/component.aaadd915.js | 6.27kb | +0.00b | 594.00ms | +120.00ms ⚠️ |
| dist/png-chunks-extract.417e691a.js | 3.55kb | +0.00b | 521.00ms | +63.00ms ⚠️ |
| dist/Modal.d8614cea.js | 3.15kb | +0.00b | 507.00ms | -109.00ms 🚀 |
| dist/16.4fc0f868.js | 2.49kb | +0.00b | 587.00ms | +147.00ms ⚠️ |
| dist/date.7c1c4784.js | 1.96kb | +0.00b | 601.00ms | +52.00ms ⚠️ |
| dist/16.5678cfe5.js | 1.86kb | +0.00b | 579.00ms | -54.00ms 🚀 |
| dist/16.fb8534c0.js | 1.79kb | +0.00b | 563.00ms | +89.00ms ⚠️ |
| dist/code.17026d12.js | 1.61kb | +0.00b | 599.00ms | +50.00ms ⚠️ |
| dist/16.85df7a6e.js | 1.51kb | +0.00b | 579.00ms | +79.00ms ⚠️ |
| dist/16.34fd12e7.js | 1.46kb | +0.00b | 563.00ms | +89.00ms ⚠️ |
| dist/16.ff3f3afd.js | 1.46kb | +0.00b | 564.00ms | +64.00ms ⚠️ |
| dist/16.55d1de3e.js | 1.45kb | +0.00b | 507.00ms | +67.00ms ⚠️ |
| dist/16.9689f28f.js | 1.44kb | +0.00b | 589.00ms | +67.00ms ⚠️ |
| dist/16.ac531af9.js | 1.40kb | +0.00b | 589.00ms | +68.00ms ⚠️ |
| dist/16.c1346bbe.js | 1.36kb | +0.00b | 589.00ms | +68.00ms ⚠️ |
| dist/16.341ad0e1.js | 1.34kb | +0.00b | 507.00ms | +67.00ms ⚠️ |
| dist/16.be69595e.js | 1.32kb | +0.00b | 589.00ms | +68.00ms ⚠️ |
| dist/16.0ef0d0ed.js | 1.31kb | +0.00b | 599.00ms | +50.00ms ⚠️ |
| dist/Modal.80b784cd.js | 1.28kb | +0.00b | 520.00ms | +63.00ms ⚠️ |
| dist/16.8015c4e6.js | 1.27kb | +0.00b | 507.00ms | +67.00ms ⚠️ |
| dist/16.1e3b9f3d.js | 1.27kb | +0.00b | 579.00ms | +79.00ms ⚠️ |
| dist/16.30604232.js | 1.26kb | +0.00b | 563.00ms | +89.00ms ⚠️ |
| dist/16.a86c2eba.js | 1.26kb | +0.00b | 579.00ms | +79.00ms ⚠️ |
| dist/16.8e47d46b.js | 1.26kb | +0.00b | 579.00ms | +79.00ms ⚠️ |
| dist/16.867d534c.js | 1.26kb | +0.00b | 563.00ms | +89.00ms ⚠️ |
| dist/component.0a97dc28.js | 1.23kb | +0.00b | 582.00ms | -33.00ms 🚀 |
| dist/action.946aff12.js | 1.23kb | +0.00b | 599.00ms | +50.00ms ⚠️ |
| dist/16.a43add3b.js | 1.23kb | +0.00b | 582.00ms | -33.00ms 🚀 |
| dist/decision.0a6d7582.js | 1.21kb | +0.00b | 611.00ms | +62.00ms ⚠️ |
| dist/16.2e0d51ab.js | 1.18kb | +0.00b | 589.00ms | +68.00ms ⚠️ |
| dist/index.html | 119.00b | +0.00b | 521.00ms | +36.00ms ⚠️ |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 16.57s | -241.00ms |
| Cached | 725.00ms | +4.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/285025750404301324)**

**Comment by mischnic** [comment]
🎉 

> mischnic added an additional internalized asset test

I did? I've just added an example (which was my playground for triggering both internalizing cases when experimenting)

File: packages/runtimes/js/src/JSRuntime.js [hunk 1]
```diff
@@ -321,7 +322,10 @@ function getLoaderRuntime({
   }
 
   if (bundle.env.outputFormat === 'global') {
-    loaderCode += `.then(() => module.bundle.root('${bundleGraph.getAssetPublicId(
+    let requireFunction = bundle.env.scopeHoist
+      ? `parcelRequire`
+      : `module.bundle.root`;
+    loaderCode += `.then(() => ${requireFunction}('${bundleGraph.getAssetPublicId(
```
Comment (mischnic) [inline packages/runtimes/js/src/JSRuntime.js:328]: I think this can be reverted as well.

File: packages/runtimes/js/src/JSRuntime.js [hunk 2]
```diff
@@ -100,17 +100,18 @@ export default (new Runtime({
...
-            './' + path.relative(options.projectRoot, resolved.value.filePath),
-          )}))`,
-          dependency,
-        });
+        if (!bundle.env.scopeHoist) {
+          // If this bundle already has the asset this dependency references,
+          // return a simple runtime of `Promise.resolve(internalRequire(assetId))`.
+          // The linker handles this for scope-hoisting.
+          assets.push({
+            filePath: path.join(options.projectRoot, 'JSRuntime.js'),
```
Comment (mischnic) [inline packages/runtimes/js/src/JSRuntime.js:108]:
```suggestion
            filePath: __filename
```

**Review by mischnic — APPROVED** [review]
Is this covered by tests? (Both cases: internalized & available in the bundle + internalized & only in parent bundle)

**Comment by wbinnssmith** [comment]
>(Both cases: internalized & available in the bundle + internalized & only in parent bundle)

I believe available in the bundle is covered here:

https://github.com/parcel-bundler/parcel/blob/a4cf3e958ea1136b4829eb93615fba5328213471/packages/core/integration-tests/test/scope-hoisting.js#L4321

...and only in the parent bundle is covered here:

https://github.com/parcel-bundler/parcel/blob/a4cf3e958ea1136b4829eb93615fba5328213471/packages/core/integration-tests/test/scope-hoisting.js#L4345

Is there another case you had in mind? Like one with another bundle in between?

**Comment by mischnic** [comment]
> Is there another case you had in mind?

No, seems like it's covered.

**Pull request merged**